src/main.rs

```diff
@@ -38,13 +52,54 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
     let addr = "0.0.0.0:4243".parse().unwrap();
     tracing::info!("chariott listening on {addr}");
 
-    let router = Server::builder()
-        .add_service(ChariottServiceServer::new(ChariottServer::new(registry, broker)));
+    let server = Arc::new(ChariottServer::new(registry, broker));
+    let router =
+        Server::builder().add_service(ChariottServiceServer::from_arc(Arc::clone(&server)));
 
     #[cfg(build = "debug")]
     let router = router.add_service(reflection_service);
 
-    router.serve_with_ctrl_c_shutdown(addr).await?;
+    let error_cancellation_token = CancellationToken::new();
+    let ctrl_c_cancellation_token = ctrl_c_cancellation();
+
+    let prune_loop = {
```

Comment:
consider moving this logic outside of the main.rs to keep it tidy.

Reply:
Done with a1b5b40955432bfd2cf05d25a3af09413098b91d.